tests/meshcop/commissioner.cpp

```diff
@@ -460,106 +420,140 @@ void HandleCommissionerSetResponse(const Coap::Message &aMessage, void *aContext
+void Commissioner::UpdateFdSet(fd_set & aReadFdSet,
+                               fd_set & aWriteFdSet,
+                               fd_set & aErrorFdSet,
+                               int &    aMaxFd,
+                               timeval &aTimeout)
+{
+    FD_SET(mSslClientFd.fd, &aReadFdSet);
+    aMaxFd = utils::Max(mSslClientFd.fd, aMaxFd);
+    FD_SET(mJoinerSessionClientFd, &aReadFdSet);
+    aMaxFd = utils::Max(mJoinerSessionClientFd, aMaxFd);
+    mJoinerSession.UpdateFdSet(aReadFdSet, aWriteFdSet, aErrorFdSet, aMaxFd, aTimeout);
+    (void)aTimeout;
+    (void)aWriteFdSet;
+    (void)aErrorFdSet;
+}
 
-    tlv->SetType(Meshcop::kCommissionerSessionId);
-    tlv->SetValue(aContext.mCommissionerSessionId);
-    otbrLog(OTBR_LOG_INFO, "COMMISSIONER_SET.req: session-id=%d", aContext.mCommissionerSessionId);
-    tlv = tlv->GetNext();
+void Commissioner::Process(const fd_set &aReadFdSet, const fd_set &aWriteFdSet, const fd_set &aErrorFdSet)
+{
+    uint8_t buffer[kSizeMaxPacket];
+    timeval nowTime;
 
-    ok = CommissionerComputeSteering();
-    /* Note: Steering computation will have logged the steering data */
-    if (!ok)
+    mJoinerSession.Process(aReadFdSet, aWriteFdSet, aErrorFdSet);
+    if (FD_ISSET(mSslClientFd.fd, &aReadFdSet))
     {
-        CommissionerUtilsFail("Cannot compute steering data\n");
-    }
-
-    tlv->SetType(Meshcop::kSteeringData);
-    tlv->SetValue(aContext.mJoiner.mSteeringData.GetDataPointer(), aContext.mJoiner.mSteeringData.GetLength());
-    tlv = tlv->GetNext();
+        int n = mbedtls_ssl_read(&mSsl, buffer, sizeof(buffer));
 
-    message->SetPath("c/cs");
-    otbrLog(OTBR_LOG_INFO, "COMMISSIONER_SET.req: coap-uri: %s", "c/cs");
-    message->SetPayload(buffer, LengthOf(buffer, tlv));
-    otbrLog(OTBR_LOG_INFO, "COMMISSIONER_SET.req: sent");
-    aContext.mCoap->Send(*message, NULL, 0, HandleCommissionerSetResponse, &aContext);
-    aContext.mCoap->FreeMessage(message);
+        if (n > 0)
+        {
+            mCoapAgent->Input(buffer, n, NULL, 0);
+        }
+    }
 
-    do
+    if (FD_ISSET(mJoinerSessionClientFd, &aReadFdSet))
     {
-        ret = mbedtls_ssl_read(aContext.mSsl, buffer, sizeof(buffer));
-        if (ret > 0)
+        struct sockaddr_in from_addr;
+        socklen_t          addrlen;
+        ssize_t            n =
+            recvfrom(mJoinerSessionClientFd, buffer, sizeof(buffer), 0, (struct sockaddr *)(&from_addr), &addrlen);
+
+        if (n > 0)
         {
-            aContext.mCoap->Input(buffer, static_cast<uint16_t>(ret), NULL, 0);
-            switch (aContext.mState)
             {
-            case kStateReady:
-                ret = 0;
-                break;
-
-            case kStateAccepted:
-                ret = MBEDTLS_ERR_SSL_WANT_READ;
-                break;
-
-            default:
-                break;
+                char buf[100];
+                GetIPString((struct sockaddr *)(&from_addr), buf, sizeof(buf));
```

Comment:
This one not very possible, since I cannot decide whether the dtls server is working on an IPV6 or IPV4 stack, but I checked some similar code and updated them wherever the IP stack can be decided.